agessh/keysource.go

```diff
@@ -0,0 +1,241 @@
+
+const (
+	fileEnv = "SOPS_AGE_SSH_PRIVATE_KEY"
+)
+
+func init() {
+	log = logging.NewLogger("AGE-SSH")
+}
+
+// MasterKey is an age key used to encrypt and decrypt sops' data key.
+type MasterKey struct {
+	Identity     string // ssh private key
+	PublicKey    string // ssh public key
+	EncryptedKey string // a sops data key encrypted with age
+
+	parsedRecipient age.Recipient // a parsed age recipient from ssh public key
+}
+
+// Encrypt takes a sops data key, encrypts it with age using the ssh public key and stores the result in the EncryptedKey field.
+func (key *MasterKey) Encrypt(datakey []byte) error {
+	buffer := &bytes.Buffer{}
+
+	if key.parsedRecipient == nil {
+		parsedRecipient, err := parseRecipient(key.PublicKey)
+
+		if err != nil {
+			log.WithField("recipient", key.parsedRecipient).Error("Encryption failed")
+			return err
+		}
+
+		key.parsedRecipient = parsedRecipient
+	}
+
+	aw := armor.NewWriter(buffer)
+	w, err := age.Encrypt(aw, key.parsedRecipient)
+	if err != nil {
+		return fmt.Errorf("failed to open file for encrypting sops data key with age ssh: %w", err)
+	}
+
+	if _, err := w.Write(datakey); err != nil {
+		log.WithField("recipient", key.parsedRecipient).Error("Encryption failed")
+		return fmt.Errorf("failed to encrypt sops data key with age: %w", err)
+	}
+
+	if err := w.Close(); err != nil {
+		log.WithField("recipient", key.parsedRecipient).Error("Encryption failed")
+		return fmt.Errorf("failed to close file for encrypting sops data key with age: %w", err)
+	}
+
+	if err := aw.Close(); err != nil {
+		log.WithField("recipient", key.parsedRecipient).Error("Encryption failed")
+		return fmt.Errorf("failed to close armored writer: %w", err)
+	}
+
+	key.EncryptedKey = buffer.String()
+
+	log.WithField("recipient", key.parsedRecipient).Info("Encryption succeeded")
+
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided sops' data key and encrypts it if it hasn't been encrypted yet.
+func (key *MasterKey) EncryptIfNeeded(datakey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(datakey)
+	}
+
+	return nil
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds.
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key.
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Decrypt decrypts the EncryptedKey field with the age identity and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	ageKeyFilePath, ok := os.LookupEnv(fileEnv)
+
+	if !ok {
+		userConfigDir, err := os.UserHomeDir()
+		if err != nil {
+			return nil, fmt.Errorf("user config directory could not be determined: %w", err)
+		}
+
+		// Check if there is a Ed25519 Key
+		ageKeyFilePath = filepath.Join(userConfigDir, ".ssh", "id_ed25519")
+		_, err = os.Stat(ageKeyFilePath)
+		if err != nil {
+			if err == os.ErrNotExist {
```

Comment:
This seems to always fail and the code never falls back to looking for `id_rsa`.
Changing the check to `if os.IsNotExist(err)` makes it work correctly.
@iamd3vil